leakpro/attacks/utils/boosting.py

```diff
@@ -0,0 +1,212 @@
+            self.memorization_threshold = 1-self.min_num_memorization_audit_points/self.org_audit_data_length
+            self.logger.info(f'Setting "memorization_threshold" to {self.memorization_threshold}')
+
+        if self.memorization_threshold != 0.0 and self.num_memorization_audit_points != 0:
+            self.logger.info("Warning!")
+            self.logger.info('"memorization_threshold" and "num_memorization_audit_points" is both used.')
+            self.logger.info('Defaulting to "num_memorization_audit_points"')
+
+        elif self.memorization_threshold != 0.0:
+            self.num_memorization_audit_points = int(self.org_audit_data_length * (1-self.memorization_threshold))
+
+        # Initialize memorization and privacy score
+        self.memorization_score = np.zeros(self.audit_data_length, dtype=float)
+        self.privacy_score = np.zeros(self.audit_data_length, dtype=float)
+
+        self._memorization_score()
+
+        if self.use_privacy_score:
+            self._privacy_score()
+
+        mem_mask, privacy_mask = self.adjust_memorization_mask()
+        mask = (mem_mask | privacy_mask)
+
+        return mask, self.memorization_score, self.privacy_score
+
+    def _memorization_score(self:Self) -> None:
+        """Run memorization score enhancement.
+
+        Memorization enhances the attack performance by only inlucing vulnerable data points
+        """
+
+        logits_function = ModelLogits()
+        logits = np.swapaxes(logits_function(self.shadow_models, self.handler, self.audit_data_indices, self.batch_size), 0, 1)
+        target_logits = np.swapaxes(logits_function([self.target_model], self.handler, self.audit_data_indices, self.batch_size),\
+                                    0, 1).squeeze()
+
+        logits = self.softmax_logits(logits)
+        target_logits = self.softmax_logits(target_logits)
+
+        if self.online:
+            for i, (logit, mask, label) in tqdm(enumerate(zip(logits, self.in_indices_mask, self.audit_data_labels)),
+                                                total=len(logits),
+                                                desc="Calculating memorization score",
+                                                leave=False
+                                                ):
+                self.memorization_score[i] = np.mean(logit[mask, label]) - np.mean(logit[~mask, label])
+        else:
+            for i, (logit, target_logit, label) in tqdm(enumerate(zip(logits, target_logits, self.audit_data_labels)),
+                                                total=len(logits),
+                                                desc="Calculating memorization score",
+                                                leave=False
+                                                ):
+                self.memorization_score[i] = np.mean(target_logit[label]) - np.mean(logit[:, label])
+
+    def _privacy_score(self:Self) -> None:
```

Comment:
So the privacy score is the same thing as the memorizations core except that we normalize by the std. Does it help to use both or do we get the same performance as with only one of them?

Reply:
I'd say they are related, the memorisation just takes the confidence of the model for the correct label while privacy score first rescales the logits using both the correct label and the incorrect ones then normalizes them with the std. One big difference here is that memorization is bounded by -1 and 1 where privacy score is not bounded.